Make it possible to have not all languages required

matthiashamacher opened the pull request:
<!--

Before working on a contribution, you must determine on which branch you need to work:
- Bug fix: choose the latest maintenance branch `11.2`
- Feature/Improvement: choose `11.x` 

> All bug fixes merged into the latest maintenance branch are also merged to the current dev branch (`11.x`) on a regular basis.

## Please make sure your PR complies with all of the following points: 
- [ ] Read and accept our [contributing guidelines](/CONTRIBUTING.md) before you submit a PR.
- [ ] Features need to be proper documented in `doc/` 
- [ ] Bugfixes need a short guide how to reproduce them -> target branch is the oldest supported maintenance branch, e.g. `11.2` (see Readme.md for the list of supported versions)
- [ ] Meet all coding standards (see PhpStan actions) 

**Don't submit a PR if it doesn't comply, it'll be closed without a comment!**
-->  
  

## Changes in this pull request  
Resolves #2958 

## Additional info
This PR adds the core functionality needed to have not all languages mandatory. For that we add a new config node `required_languages` to the `pimcore.general` node, which stores similar to the `valid_languages` all required languages in an array. This data is then also made available to the frontend via the `SystemSettingsConfig::prepareSystemConfig()` function.
In the `checkValidity()` function of the Localizedfield we then not check all languages but only the required ones

matthiashamacher committed:
Make it possible to have not all languages required

Signed-off-by: Matthias Hamacher <m.hamacher@basecom.de>

matthiashamacher committed:
Update documentation

Signed-off-by: Matthias Hamacher <m.hamacher@basecom.de>

mcop1 commented:
@matthiashamacher nice addition, thank you! Could you please have a look at the conflict?

blankse commented on bundles/CoreBundle/src/DependencyInjection/Configuration.php:
```diff
@@ -222,6 +222,15 @@ private function addGeneralNode(ArrayNodeDefinition $rootNode): void
+                    ->ifString()
+                        ->then(fn ($v) => explode(',', $v))
+                    ->end()
+                    ->defaultValue(['en', 'de', 'fr'])
```
I think the default value should be empty and a empty array should make all valid languages required. Otherwise we have a behavior change after an update, right?

mcop1 replied:
Very good point, yes. Maybe it would already be enough to just return `Tool::getValidLanguages()` from within `Tool::getRequiredLanguages()` instead of `return []`, once the `defaultValue` in the config gets removed.

matthiashamacher replied:
Good Point. Thought setting the same value as the valid languages would be enough to not have problems after the update, but you are right this could have caused issues. I liked the idea to just use the `Tool::getValidLangauages()`  function as fallback, to have the same languages and therefore make all required. Having an empty array as all would have made everything else slightly different, so I thought this also would be the better solution.

matthiashamacher committed:
Merge branch '11.x' into feature/one-required-language

Signed-off-by: Matthias Hamacher <m.hamacher@basecom.de>

# Conflicts:
#	lib/SystemSettingsConfig.php

matthiashamacher committed:
Fix behavior change

Remove the default value of the config and returning all valid languages if there are no required languages set

Signed-off-by: Matthias Hamacher <m.hamacher@basecom.de>

mcop1 commented:
Thanks again, really nice addition :100:
